### PR TITLE
Serverless new cr

### PR DESCRIPTION
**Pull request by dbadura**
**Description**

Changes proposed in this pull request:

- Proposal for new serverless cr

**Related issue(s)**

See: https://github.com/kyma-project/kyma/issues/12889

### CODE REVIEW

File: components/serverless/api/v1beta1/function_types.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+type Source struct {
+	// +optional
+	ArtifactSource *ArtifactSource `json:"artifactSource"`
+	//TODO: we should check if it is possible
+	// +optional
+	CRDSource *CRDSource `json:"crdSource"`
+	// +optional
+	GitRepositorySource *GitRepositorySource `json:"gitRepositorySource"`
+	// +optional
+	InlineSource *InlineSource `json:"inlineSource"`
```
Comment (moelsayed) [inline components/serverless/api/v1beta1/function_types.go:43]: I think the `Source` suffix is redundant there.

File: components/serverless/api/v1beta1/function_types.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+type CRDSource struct {
+	Api        string `json:"api"`
+	Version    string `json:"version"`
+	Kind       string `json:"kind"`
+	Name       string `json:"name"`
+	Namespaces string `json:"namespaces"`
```
Comment (moelsayed) [inline components/serverless/api/v1beta1/function_types.go:61]: Do you thing it makes sense to use `ObjectMeta` and `TypeMeta` for these fields instead?

File: components/serverless/api/v1beta1/function_types.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+type ArtifactSource struct {
+	//TODO: maybe it's worth do distinguish between local and external, as we know the internal URL
+	URL     string `json:"url"`
+	BaseDir string `json:"baseDir"`
+	// +optional
+	Credentials *Credentials `json:"credentials"`
+}
+
+type ConfigMapSelector struct {
```
Comment (moelsayed) [inline components/serverless/api/v1beta1/function_types.go:76]: I don't think `selector` is the correct term here?

File: components/serverless/api/v1beta1/function_types.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	Password v1.SecretKeySelector `json:"password"`
+}
+
+type Oauth struct {
+	OauthURL string               `json:"oatuhURL"`
+	User     v1.SecretKeySelector `json:"user"`
+	Password v1.SecretKeySelector `json:"password"`
+}
+
+type Template struct {
```
Comment (moelsayed) [inline components/serverless/api/v1beta1/function_types.go:105]: What is this for?

File: components/serverless/api/v1beta1/function_types.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	RuntimeProfile *string `json:"runtimeProfile"`
+
+	// +optional
+	Resources *v1.ResourceRequirements `json:"resources,omitempty"`
+
+	// +optional
+	BuildProfile *string `json:"buildProfile"`
+
+	// +optional
+	BuildResources *v1.ResourceRequirements `json:"buildResources,omitempty"`
```
Comment (moelsayed) [inline components/serverless/api/v1beta1/function_types.go:138]: How do these fields work together?

**Review by moelsayed — COMMENTED** [review]
Just a a few questions

File: components/serverless/api/v1beta1/function_types.go (hunk 3 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go:76]: You're right, it should be Ref.

File: components/serverless/api/v1beta1/function_types.go (hunk 4 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go:105]: This is for defining annotations and labels on deployment of function.

File: components/serverless/api/v1beta1/function_types.go (hunk 2 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go:61]: Can those objects be used to reference particular crd?
I looked at them and they carry many information which I think we don't need.

File: components/serverless/api/v1beta1/function_types.go (hunk 5 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go:138]: I wanted to make it similar to the [k8s volume source](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L49)
Generally you have to choose only one of them.

Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go:138]: You can choose predefined profile or create custom one.

File: components/serverless/api/v1beta1/function_types.go (hunk 1 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go]: Removed

File: components/serverless/api/v1beta1/function_types.go (hunk 2 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go:61]: Could you give me a example of what those 2 objects expose and can be helpful?

File: components/serverless/api/v1beta1/function_types.go (hunk 5 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go]: What about this:
```
build:
 - profile
 - resources
```
```
runtime:
 - profile
 - resources
```
And only one of those options should be filled.

Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go]: New idea
```
requirements:
 -build:
    -profile
    -resources
 -runtime
  - profile
  - resources
  ```

Comment (moelsayed) [reply components/serverless/api/v1beta1/function_types.go]: I like the latest idea more. And you are right about the `runtime` name. I would also suggest `deployment` or `instance` for this.

File: components/serverless/api/v1beta1/function_types.go (hunk 2 above)
Comment (moelsayed) [reply components/serverless/api/v1beta1/function_types.go:61]: Nothing comes to my mind at this point.

File: components/serverless/api/v1beta1/function_types.go (hunk 5 above)
Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go]: What do you think about `function`?
```
requirements/resourceConfig:
 -build:
    -profile
    -resources
 -function
  - profile
  - resources
  ```

Comment (dbadura) [reply components/serverless/api/v1beta1/function_types.go]: I've updated the cr.

**Pull request closed without merging**